File: ios/MenuMaker/Data/Repositories/AuthRepository.swift
import Foundation
import Combine

/// Authentication repository
@MainActor
class AuthRepository: ObservableObject {
    static let shared = AuthRepository()

    private let apiClient = APIClient.shared
    private let keychainManager = KeychainManager.shared

    private init() {}

    // MARK: - Authentication

    func login(email: String, password: String) async throws -> AuthData {
        let request = LoginRequest(email: email, password: password)

        let response: AuthResponse = try await apiClient.request(
            endpoint: AppConstants.API.Endpoints.login,
            method: .post,
            body: request,
            requiresAuth: false
        )

        // Save tokens
        try await keychainManager.saveToken(response.data.accessToken)
        try await keychainManager.saveRefreshToken(response.data.refreshToken)
        try await keychainManager.saveUserId(response.data.user.id)
        try await keychainManager.saveUserEmail(response.data.user.email)

        return response.data
    }

    func signup(email: String, password: String, name: String, phone: String? = nil) async throws -> AuthData {
        let request = SignupRequest(email: email, password: password, name: name, phone: phone)

        let response: AuthResponse = try await apiClient.request(
            endpoint: AppConstants.API.Endpoints.signup,
            method: .post,
            body: request,
            requiresAuth: false
        )

        // Save tokens
        try await keychainManager.saveToken(response.data.accessToken)
        try await keychainManager.saveRefreshToken(response.data.refreshToken)
        try await keychainManager.saveUserId(response.data.user.id)
        try await keychainManager.saveUserEmail(response.data.user.email)

        return response.data
    }

    func logout() async throws {
        do {
            // Try to logout on server
            let _: EmptyResponse = try await apiClient.request(
                endpoint: AppConstants.API.Endpoints.logout,
                method: .post
            )
        } catch {
            // Continue even if server logout fails
            print("Server logout failed: \(error)")
        }

        // Always clear local tokens
        await keychainManager.clearTokens()
    }

    func getCurrentUser() async throws -> User {
        let response: AuthResponse = try await apiClient.request(
            endpoint: AppConstants.API.Endpoints.me,
            method: .get
        )

        return response.data.user
    }

    func refreshSession() async throws {
        _ = try await getCurrentUser()
    }

    // MARK: - Session Check

    func isAuthenticated() async -> Bool {
        do {
            let token = try await keychainManager.getToken()
            return token != nil && !token!.isEmpty
        } catch {
            return false
        }
    }

    func getUserId() async throws -> String? {
        try await keychainManager.getUserId()
    }

    func getUserEmail() async throws -> String? {
        try await keychainManager.getUserEmail()
    }

    // MARK: - Password Reset

    func sendPasswordReset(email: String) async throws {
        let request = ForgotPasswordRequest(email: email)

        let _: EmptyResponse = try await apiClient.request(
            endpoint: AppConstants.API.Endpoints.forgotPassword,
            method: .post,
            body: request,
            requiresAuth: false
        )
    }
}

// MARK: - Empty Response

struct EmptyResponse: Decodable {
    let success: Bool
}

File: android/app/src/main/kotlin/com/menumaker/data/repository/AuthRepository.kt
package com.menumaker.data.repository

import com.menumaker.data.common.Resource
import com.menumaker.data.local.datastore.TokenDataStore
import com.menumaker.data.remote.api.ApiService
import com.menumaker.data.remote.models.AuthData
import com.menumaker.data.remote.models.LoginRequest
import com.menumaker.data.remote.models.SignupRequest
import com.menumaker.data.remote.models.UserDto
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject

interface AuthRepository {
    fun login(email: String, password: String): Flow<Resource<AuthData>>
    fun signup(email: String, password: String, name: String): Flow<Resource<AuthData>>
    fun logout(): Flow<Resource<Unit>>
    fun isAuthenticated(): Flow<Boolean>
    fun sendPasswordReset(email: String): Flow<Resource<Unit>>
    fun updateProfile(updates: Map<String, Any>): Flow<Resource<AuthData>>
    fun changePassword(current: String, new: String): Flow<Resource<Unit>>
    fun getCurrentUser(): Flow<Resource<UserDto>>
}

class AuthRepositoryImpl @Inject constructor(
    private val apiService: ApiService,
    private val tokenDataStore: TokenDataStore
) : AuthRepository {

    override fun login(email: String, password: String): Flow<Resource<AuthData>> = flow {
        emit(Resource.Loading)
        try {
            val response = apiService.login(LoginRequest(email, password))
            if (response.isSuccessful && response.body() != null) {
                val authData = response.body()!!.data
                tokenDataStore.saveTokens(authData.accessToken, authData.refreshToken)
                tokenDataStore.saveUserId(authData.user.id)
                tokenDataStore.saveUserEmail(authData.user.email)
                emit(Resource.Success(authData))
            } else {
                emit(Resource.Error(response.message() ?: "Login failed"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }

    override fun signup(email: String, password: String, name: String): Flow<Resource<AuthData>> = flow {
        emit(Resource.Loading)
        try {
            val response = apiService.signup(SignupRequest(email, password, name))
            if (response.isSuccessful && response.body() != null) {
                val authData = response.body()!!.data
                tokenDataStore.saveTokens(authData.accessToken, authData.refreshToken)
                tokenDataStore.saveUserId(authData.user.id)
                tokenDataStore.saveUserEmail(authData.user.email)
                emit(Resource.Success(authData))
            } else {
                emit(Resource.Error(response.message() ?: "Signup failed"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }

    override fun logout(): Flow<Resource<Unit>> = flow {
        emit(Resource.Loading)
        try {
            apiService.logout()
            tokenDataStore.clearTokens()
            emit(Resource.Success(Unit))
        } catch (e: Exception) {
            // Still clear local tokens even if API call fails
            tokenDataStore.clearTokens()
            emit(Resource.Success(Unit))
        }
    }

    override fun isAuthenticated(): Flow<Boolean> = flow {
        tokenDataStore.getAccessToken().collect { token ->
            emit(!token.isNullOrEmpty())
        }
    }

    override fun sendPasswordReset(email: String): Flow<Resource<Unit>> = flow {
        emit(Resource.Loading)
        try {
            val response = apiService.sendPasswordReset(mapOf("email" to email))
            if (response.isSuccessful) {
                emit(Resource.Success(Unit))
            } else {
                emit(Resource.Error(response.message() ?: "Failed to send password reset email"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }

    override fun updateProfile(updates: Map<String, Any>): Flow<Resource<AuthData>> = flow {
        emit(Resource.Loading)
        try {
            val response = apiService.updateUserProfile(updates)
            if (response.isSuccessful && response.body() != null) {
                val authData = response.body()!!.data
                tokenDataStore.saveTokens(authData.accessToken, authData.refreshToken)
                tokenDataStore.saveUserId(authData.user.id)
                tokenDataStore.saveUserEmail(authData.user.email)
                emit(Resource.Success(authData))
            } else {
                emit(Resource.Error(response.message() ?: "Failed to update profile"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }

    override fun changePassword(current: String, new: String): Flow<Resource<Unit>> = flow {
        emit(Resource.Loading)
        try {
            val request = mapOf("current_password" to current, "new_password" to new)
            val response = apiService.changePassword(request)
            if (response.isSuccessful) {
                emit(Resource.Success(Unit))
            } else {
                emit(Resource.Error(response.message() ?: "Failed to change password"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }

    override fun getCurrentUser(): Flow<Resource<UserDto>> = flow {
        emit(Resource.Loading)
        try {
            val response = apiService.getCurrentUser()
            if (response.isSuccessful && response.body() != null) {
                val user = response.body()!!.data.user
                tokenDataStore.saveUserId(user.id)
                tokenDataStore.saveUserEmail(user.email)
                emit(Resource.Success(user))
            } else {
                emit(Resource.Error(response.message() ?: "Failed to load user"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }
}
